[Security Solution] Tines connector

semd opened the pull request:
## Summary

Issue: https://github.com/elastic/kibana/issues/140066
Doc: https://docs.google.com/document/d/14BY-6CIin1CUH5bwJJgfrGl37hWO-CeNMdl_35agpvk/edit?usp=sharing

Create a new connector type that offers low friction/low effort approach to augmenting Elastic capabilities with SOAR capabilities of Tines.

## Implementation

Tines connector implements subActionConnector. With 4 subActions configured:

- **stories**: Retrieves the User available Story objects from Tines, to render the Story selector options in the params form. It uses the `email` and `token` authentication headers from the configuration. 
It is requested only when the form opens and when the connector instance changes.

- **webhooks**: Retrieves the Story available Webhooks objects from Tines, to render the Webhook selector in the params form. It uses the `email` and `token` authentication headers from the configuration and the `story_id` parameter. 
There is no filter for `type` in the actions (a.k.a. agents) endpoint, so we have to request all actions and filter them by `type === 'Agents::WebhookAgent'` on our side.
It is requested every time the selected story changes.

- **run**: The main action execution. It sends the alerts to the Tines configured webhook, using webhook' `path` and `secret` values. There's no template to render, the data coming from the execution is just pruned (the `kibana` entry is removed from all `context.alerts`) and sent directly using the same format to Tines.

- **test**: The test form execution. It ends up calling **run** but using a parametrized body.

### Pagination
Both **stories** and **webhooks** subActions need pagination, since Tines do not expose any search endpoint for them. The current hard limit is 100 pages. The `paginatedRequest` function in the connector implementation encapsulates this logic.

## Testing

1- Create a [Tines](https://www.tines.com/) free account.

2- Create a [new Story](https://www.tines.com/docs/quickstart/simple-story) and attach a [Webhook Action](https://www.tines.com/docs/quickstart/creating-an-action) to start receiving events.

3- Create an [API token](https://www.tines.com/api/authentication)

4- Configure the Tines Connector in Kibana using the Tines tenant URL that has been generated in the Tines app, the email used to sign in, and the API token generated. [docs](https://github.com/semd/kibana/blob/140066_tines_connector/docs/management/connectors/action-types/tines.asciidoc#connector-configuration)

5- Attach the Tines Connector to a Detection Rule, selecting the Story and Webhooks created. [docs](https://github.com/semd/kibana/blob/140066_tines_connector/docs/management/connectors/action-types/tines.asciidoc#actions)

6- After each rule execution, events should appear in the Tines webhook action.

## Screenshots

Configure a Tines connector

![tines_connector_selection](https://user-images.githubusercontent.com/17747913/196389019-820aff49-6ad6-442e-a69f-3c782cbd65e6.png)

![tines_connector_config](https://user-images.githubusercontent.com/17747913/198035138-e7f3bb25-ebd1-4cfd-9cc5-b0bfe434c25c.png)

Use the Tines connector 

![tines_rule_action](https://user-images.githubusercontent.com/17747913/196389010-c87045a4-2b74-4903-9a81-ccbcff09fbf1.png)

![tine_params_form](https://user-images.githubusercontent.com/17747913/198034501-7e9ad912-111e-48b6-8387-fcf6f0663511.png)

Tines events

![tines_events](https://user-images.githubusercontent.com/17747913/196734338-91e1a397-2d03-4ee6-8ad2-16cb39abe9bf.png)

### Checklist

Delete any items that are not applicable to this PR.

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [x] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios(https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))

elasticmachine commented:
Pinging @elastic/security-threat-hunting (Team:Threat Hunting)

semd commented:
@elasticmachine merge upstream

semd commented:
@elasticmachine merge upstream

semd commented:
@elasticmachine merge upstream

semd commented:
Functional tests for the connector are still missing, but they are coming.

jonathan-buttner commented on x-pack/plugins/stack_connectors/server/connector_types/security/tines/tines.ts:
```diff
@@ -0,0 +1,189 @@
+    return result;
+  }
+
+  protected getResponseErrorMessage(error: AxiosError): string {
```
The framework intends for this method to return a string rather than throw an error. Could we remove the throwing of errors here?

jonathan-buttner reviewed (commented):
Looking good, left a few questions and change requests.

jonathan-buttner commented:
> Functional tests for the connector are still missing, but they are coming.

Could you add a few integration tests under `x-pack/test/alerting_api_integration/security_and_spaces/group2/tests/actions/connector_types/`?

jonathan-buttner commented on docs/management/connectors/action-types/tines.asciidoc:
```diff
@@ -0,0 +1,56 @@
+[role="xpack"]
```
It seems like the other connectors have a section about how to create a pre-configured one, should we add that here too?

semd commented on x-pack/plugins/stack_connectors/public/connector_types/security/index.ts:
```diff
@@ -4,3 +4,5 @@
  * 2.0; you may not use this file except in compliance with the Elastic License
  * 2.0.
  */
+
```
This connector is only for Security by now, I saw the directory was already created and I guessed it was intended for this kind of connector. If it is not the case I can move everything to `stack`.

ymao1 replied:
@jonathan-buttner Sorry for the late reply! Yes, I added this as a placeholder since I was told there would be security specific connector types but it sounds like everything will be moving to back to one location eventually anyway :)

semd commented:
> * we've created a new "grouping" of connector types named `security`. Do we really need that?  Is this connector supposed to be limited for use with a select set of rule types?  I was able to add it to an index threshold rule (but not save it, as I don't yet have a Tines account).

This connector is meant to work with Security, non-security use cases are not required by now. Let's discuss the need for the "security" directory [here](https://github.com/elastic/kibana/pull/143505#discussion_r1005732018)

> * with the sub-commands, I think that makes it not usable as a typical connector; or I'm not sure what happens (default sub-command when invoking the action?)

Not sure to understand what you mean by "sub-commands", If you are referring to the `subActionsFramework`, it is an effort from the cases team to have an easier way to create connectors, the `registerSubActionConnectorType` works as an adapter to the "typical connector" logic. There's no default sub-action (sub-command) if that's what you mean, the sub-action identifier is always required. When the action (execution) is triggered the "run" sub-action is invoked. 

> * it sounds like it's sending `context.alerts` to the service, but not all rules populate `context.alerts`, I think, though most do.  It would be more "flexible" to allow the user to select the content with their own template, and you could make the default template `{{context.alerts}}` but I believe that's not possible due to [issues with rendering JSON arrays in mustache templates](https://github.com/elastic/kibana/issues/84217).  Perhaps we can leave it like this for now, and then when we get the JSON issue resolved, we could open this up as a new param, defaulting to the existing "hardcoded" value

Yes, we have removed the body template for that reason, having JSON was a requirement, but customizing the payload format was not. Mustache syntax was only adding complexity to the connector configuration.
And yes, the idea is to try using the new JSON format templating/mapping when we have it, to send only the data the recipient need.
Currently, the connector sends everything the rule execution gives it, which is the most flexible way from the Tines point of view, they will be able to take any piece of data they need. If the `context.alerts` is not populated in some rules it won't be sent, so the Tines story attached should take care of it. But they would have the same problem using a body template.

> * Does this require some kind of "app" on the Tines side to process this info?  Like the way xMatters, ServiceNow, etc require you to "install" someone in the integration app to "process" the incoming data.   Or do the stories/webhook queries handle this aspect?  Might want to check with the Cases folks in terms of how to make error scenarios around this kind of stuff debuggable for the user, and for us :-)

It does not require any app installed on the Tines side. You can create a new profile (free tier), create a new blank Story, and attach a so-called "Webhook" action to it, they are the third-party data entry points. You can also load a pre-defined Elastic story directly from the Story Library. 
Then you will have to create an API token, from the Tines profile, to configure the connector in Security, so we can retrieve the stories/webhooks data to choose from in the connector parameters form. This is explained in this PR description.
Regarding the errors, I have asked the Tines team for the API error reference, They haven't answered yet.

semd commented on x-pack/plugins/stack_connectors/server/connector_types/security/tines/render.ts:
```diff
@@ -0,0 +1,33 @@
+
+  let body: string;
+  try {
+    body = JSON.stringify(set('context.alerts', alerts, variables));
```
Good point, the standard lodash `set` function does mutate the target, but we use the `set` from `lodash/fp` here, which does not mutate the target, it creates an entirely new `variables` object when it is called. I used it instead of doing the deep clone manually.

semd commented on x-pack/plugins/stack_connectors/server/connector_types/security/tines/tines.ts:
```diff
@@ -0,0 +1,189 @@
+  }
+
+  /**
+   * Recursively request pages for given a Tines request, it stops when there's no more pages
```
yup, I changed the implementation I forgot the comment, I'll update it.

semd commented on x-pack/plugins/stack_connectors/server/connector_types/security/tines/tines.ts:
```diff
@@ -0,0 +1,189 @@
+   * - `pageSize` the number of results per page. Optional, defaults to `20`.
+   * @returns An array of items of the response schema type
+   */
+  private async paginatedRequest<R extends TinesBaseApiResponse, T>(
```
I already asked them, and they liked the idea and they said they will study it. I don't have any other news about it though. I will ping them.

semd commented on x-pack/plugins/stack_connectors/server/connector_types/security/tines/index.ts:
```diff
@@ -0,0 +1,38 @@
+    secrets: TinesSecretsSchema,
+  },
+  validators: [{ type: ValidatorType.CONFIG, validator: urlAllowListValidator('url') }],
+  supportedFeatureIds: [AlertingConnectorFeatureId, SecurityConnectorFeatureId],
```
As we discussed, not needed for now.

semd commented on x-pack/plugins/stack_connectors/public/connector_types/security/tines/tines_params.tsx:
```diff
@@ -0,0 +1,245 @@
+      <EuiFlexItem>
+        <EuiFormRow
+          fullWidth
+          error={errors.story}
```
You're right, I missed that

semd commented on x-pack/plugins/stack_connectors/public/connector_types/security/tines/tines.ts:
```diff
@@ -0,0 +1,81 @@
+      if (!subActionParams?.webhook?.storyId) {
+        errors.subActionParams.push(translations.STORY_REQUIRED);
+      }
+      if (!subActionParams?.webhook?.id) {
```
Fixed! good catch

pmuellr commented:
> Not sure to understand what you mean by "sub-commands", If you are referring to the `subActionsFramework`, ...

Ah, forgot how the sub-action stuff worked, thanks for reminding me :-)

> It does not require any app installed on the Tines side. ... You can also load a pre-defined Elastic story directly from the Story Library. ...

I think having a sample in the store would be super-useful, and should be mentioned in the doc.  I did a search on the Tines site but didn't seen anything when searching for "elastic".

semd commented:
@pmuellr 
> I think having a sample in the store would be super-useful, and should be mentioned in the doc. I did a search on the Tines site but didn't seen anything when searching for "elastic".

I think they disappeared during these last days, but I just checked now and they are in the library again:
![tines_elastic_stories](https://user-images.githubusercontent.com/17747913/198682909-81e6668b-8286-4c17-b5e7-62654c5c756a.png)

Yes, sounds good, I will mention the library in the doc 👍

pmuellr commented on x-pack/plugins/stack_connectors/server/connector_types/security/tines/api_schema.ts:
```diff
@@ -0,0 +1,62 @@
+      {
+        next_page: schema.nullable(schema.string()),
+      },
+      { unknowns: 'ignore' }
```
It doesn't appear we use this option for other connector schemas.  I think I'd advise against it.  

With it set this way, a customer may pass some bogus property in their params/secrets/etc, and today we'd ignore it.  If we later add more properties to our schema, it's possible they were already using that property name, and the connector could break when it runs.  A type of migration failure.  We had a similar situation a while back, required a migration to fix it.

pmuellr replied:
hmmm ... I just realized these are schemas for the Tine's HTTP requests/responses, so ignore that comment ^^^ :-)

semd replied:
yes exactly, these are API response schemas, they are a bit wider to prevent throwing errors if there's more information than needed, everything should keep working.

jonathan-buttner commented on docs/management/connectors/action-types/tines.asciidoc:
```diff
@@ -45,6 +70,18 @@ image::../images/tines-alerting.png[Tines rule alert]
+=== Tines Story Libary
+
+In order to simplify the integration with Elastic, Tines offers a set of pre-defined Elastic stories in the Story library.
+They can be found by search for "Elastic" in the Tines Story library:
```
```suggestion
They can be found by searching for "Elastic" in the Tines Story library:
```

jonathan-buttner reviewed (commented):
Integration tests look great and thanks for making the other changes. Left one suggestion on the docs.

One other testing request 😬 would it be feasible to create some end to end UI tests? The tests live here (Opsgenie): https://github.com/elastic/kibana/blob/main/x-pack/test/functional_with_es_ssl/apps/triggers_actions_ui/connectors/opsgenie.ts

Just a heads up that my opsgenie PR is going to refactor those tests so if you could put yours in a separate file that'd be super helpful to avoid conflicts.

I haven't tried using a simulator in the end to end tests but I wonder if you'll be able to test much in the UI without a simulator 🤔 since it needs to make requests.

semd commented:
@elasticmachine merge upstream

semd commented:
@elasticmachine merge upstream

semd commented on x-pack/plugins/stack_connectors/public/connector_types/security/tines/tines_params.tsx:
```diff
@@ -0,0 +1,248 @@
+
+type StoryOption = EuiComboBoxOptionOption<TinesStoryObject>;
+type WebhookOption = EuiComboBoxOptionOption<TinesWebhookObject>;
+
```
requested by Tines: add the `(draft)` flag for non-published stories

semd commented on x-pack/plugins/stack_connectors/public/connector_types/security/tines/tines_params.tsx:
```diff
@@ -0,0 +1,330 @@
+    </EuiFlexItem>
+    {value?.published && (
+      <EuiFlexItem grow={false}>
+        <EuiBadge color="hollow">{i18n.STORY_PUBLISHED_BADGE_TEXT}</EuiBadge>
```
The `Published` badge was requested by Tines:

![story_published_badge](https://user-images.githubusercontent.com/17747913/200539511-1e29fb9f-a278-46b1-b846-9c809402dca4.png)

semd commented:
Implemented the alternative parameter for the scenario in which we hit the 500 results limit in the Tines API.
Tines confirmed to us that no current user will hit this limit. However, we've implemented this fallback in case it is hit in the future. It works this way:

- If any of the story or webhook requests hit the 500 results limit (`pages > 1`), the callout message and the webhook URL input are displayed.
- Users will still be able to select the story and webhook from the 500 loaded options in the selectors.
- If the user pastes the webhook URL in the form, the selectors above will be disabled. We'll use the webhook URL directly in the execution instead of generating the URL from the selected webhook. (Tines has a click-copy button for the webhook URL in the Tines app, the user will have to have access to the story.)
- If no limit was hit but the webhook URL is stored from a previous setup, the callout message won't be displayed but the webhook URL input will be shown with the saved value. The selectors will still be disabled.
- If the user removes the webhook URL, the story and webhook selectors will become active again.

It has been approved by Security UX and Product.

![webhookUrl](https://user-images.githubusercontent.com/17747913/200532483-f25c9f5b-9904-4721-b32d-a622b19950ac.png)
cc @jonathan-buttner @cnasikas @XavierM

cnasikas commented on docs/management/action-types.asciidoc:
```diff
@@ -59,6 +59,10 @@ a| <<swimlane-action-type,{swimlane}>>
 
 | Create an incident in {swimlane}.
 
+a| <<tines-action-type,Tines>>
```
@lcawl @mikecote Do we list all of the connectors regardless of where they are supported (stack, security solution, o11y, cases, etc)?

mikecote replied:
Yes, we've been keeping this page as the source of truth for the connector documentation.

cnasikas commented on x-pack/plugins/stack_connectors/server/connector_types/security/tines/index.ts:
```diff
@@ -0,0 +1,38 @@
+    secrets: TinesSecretsSchema,
+  },
+  validators: [{ type: ValidatorType.CONFIG, validator: urlAllowListValidator('url') }],
+  supportedFeatureIds: [SecurityConnectorFeatureId],
```
@ymao1 Should not it show "Security solution"? 

<img width="1448" alt="Screenshot 2022-11-08 at 7 52 21 PM" src="https://user-images.githubusercontent.com/7871006/200642221-7f50fb73-b5bc-4168-a04f-73bd3921a003.png">

ymao1 replied:
No, we are showing the feature ID category now, which is either `Alerting Rules` or `Cases`. Since this is available for Security rules, that falls under `Alerting Rules`

cnasikas replied:
Oh, I was not aware of that. Thanks for clarifying!

cnasikas commented on x-pack/test/alerting_api_integration/security_and_spaces/group2/tests/actions/connector_types/security/tines.ts:
```diff
@@ -0,0 +1,557 @@
+              data: tinesWebhookSuccessResponse,
+            });
+          });
+        });
```
We should add a test for testing the `test` sub action.

cnasikas reviewed (commented):
Very good job with the connector @semd! I left some comments. I tested without any issues. I noticed that when the user first enters to the params form it show an error. This is a bit misleading because the user did not interact with the form yet. Other connectors do not show an error until the user interacts with the form.

<img width="1032" alt="Screenshot 2022-11-08 at 7 53 41 PM" src="https://user-images.githubusercontent.com/7871006/200825592-6cd28fc2-4aeb-4198-ae59-314fd05e10ce.png">
<img width="1021" alt="Screenshot 2022-11-08 at 7 53 34 PM" src="https://user-images.githubusercontent.com/7871006/200825598-599d4abc-c8ee-4658-bbbf-e23f57447e13.png">

mikecote commented:
FYI I just merged https://github.com/elastic/kibana/pull/144736 and this PR will require to add tines into the list within `x-pack/test/alerting_api_integration/spaces_only/tests/actions/check_registered_connector_types.ts` after pulling latest.

semd commented on x-pack/plugins/triggers_actions_ui/public/application/hooks/use_sub_action.tsx:
```diff
@@ -117,11 +126,11 @@ export const useSubAction = <P, R>({
-      } catch (e) {
-        if (isMounted && !abortCtrl.signal.aborted) {
+      } catch (err) {
+        if (isActive) {
```
before we abort at line 143 we set `isActive = false`, it prevents the error to be dispatched.

semd commented on x-pack/plugins/stack_connectors/server/connector_types/security/tines/tines.ts:
```diff
@@ -0,0 +1,175 @@
+  private async tinesApiRequest<R extends TinesBaseApiResponse, T>(
+    req: SubActionRequestParams<R>,
+    reducer: (response: R) => T,
+    pageSize: number = 500 // maximum page size allowed
```
I'll remove the parameter since it is not used, and move the `500` to a constant

jonathan-buttner commented on x-pack/plugins/stack_connectors/server/connector_types/security/tines/render.ts:
```diff
@@ -0,0 +1,33 @@
+
+  let body: string;
+  try {
+    body = JSON.stringify(set('context.alerts', alerts, variables));
```
Just confirming that we want to send the entire `variables` field as the body and not jus the `alerts`?

![image](https://user-images.githubusercontent.com/56361221/200944014-6a4cdecb-2f6a-4a76-922c-a6b6622c767b.png)

cnasikas replied:
Also, https://github.com/elastic/kibana/pull/143505#discussion_r1017819185

semd replied:
yes, this is the expected payload. we confirmed that with Tines team. 
It is the entire `variables` body with one small nuance: we clean the `kibana` entry inside the alert data, if present, to reduce some weight since this entry has a lot of duplicated information (for all items in the _context.alerts_ array).
Once the framework has an alternative to Mustache syntax, and we are able to transform/map the execution content to JSON format, instead of ndJSON, we can start using it in this connector.

jonathan-buttner reviewed (commented):
Left a few questions. Tested the new changes and things look good though!

stephmilovic commented on x-pack/plugins/stack_connectors/public/connector_types/security/tines/tines_connector.test.tsx:
```diff
@@ -0,0 +1,159 @@
+        </ConnectorFormTestProvider>
+      );
+
+      await act(async () => {
```
I see without this, we get the `Warning: An update to UseFieldComp inside a test was not wrapped in act(...).` error

cnasikas replied:
The reason you eliminate this warning is not because of `act` but because of `async/await`. You can read more about it here: https://kentcdodds.com/blog/fix-the-not-wrapped-in-act-warning.

stephmilovic commented on x-pack/plugins/stack_connectors/public/connector_types/security/tines/translations.ts:
```diff
@@ -0,0 +1,262 @@
+  'xpack.stackConnectors.security.tines.params.componentWarning.webhookNotFound',
+  {
+    defaultMessage:
+      'Can not find the saved webhook. Please select a valid webhook from the selector',
```
```suggestion
      'Cannot find the saved webhook. Please select a valid webhook from the selector',
```

stephmilovic commented on x-pack/plugins/stack_connectors/public/connector_types/security/tines/translations.ts:
```diff
@@ -0,0 +1,262 @@
+    defaultMessage: 'Tines API results limit reached',
+  }
+);
+export const WEBHOOK_URL_FALLBACK_TEXT = (type: 'Story' | 'Webhook') =>
```
I think typescript will be angry for using `type` because its a reserved word
```suggestion
export const WEBHOOK_URL_FALLBACK_TEXT = (tinesType: 'Story' | 'Webhook') =>
```

stephmilovic commented on x-pack/test/alerting_api_integration/spaces_only/tests/actions/check_registered_connector_types.ts:
```diff
@@ -42,6 +42,7 @@ export default function createRegisteredConnectorTypeTests({ getService }: FtrPr
         '.jira',
         '.resilient',
         '.teams',
+        '.tines',
```
I see your test is failing because the array is in the wrong order. i think you should add a `.sort()` to these arrays so the order does not matter

stephmilovic reviewed (approved):
Reviewed the code and tested locally, all works as expected. This is a very interesting system, good job navigating it and writing this solution. Thanks @semd!

LGTM from Explore team!

YulNaumenko reviewed (approved):
Thank you a lot for such a great effort! Tested locally, connector works as described in the document. Code LGTM!

cnasikas commented on x-pack/plugins/stack_connectors/server/connector_types/security/tines/render.ts:
```diff
@@ -0,0 +1,33 @@
+  try {
+    body = JSON.stringify(set('context.alerts', alerts, variables));
+  } catch (err) {
+    body = `error generating body payload: ${err.message}`;
```
If this is the case, then I think it is better to standardize the error schema we sent to be future-proof. What about `{ error: { message } }` ?

semd commented on x-pack/test/alerting_api_integration/common/fixtures/plugins/actions_simulators/server/tines_simulation.ts:
```diff
@@ -0,0 +1,155 @@
+  }
+}
+
+export function initPlugin(router: IRouter, path: string) {
```
yes, this simulator is used by 2 tests: this _initPlugin_ export is used by `x-pack/test/functional_with_es_ssl/apps/triggers_actions_ui/connectors/tines.ts` and the _TinesSimulator_ class is used by `x-pack/test/alerting_api_integration/security_and_spaces/group2/tests/actions/connector_types/security/tines.ts`. 

I followed the same pattern as other existing simulators, some only export the initPlugin function, and others only export the class (Opsgenie). This is the first one to be used for both kinds of tests.

semd commented on x-pack/plugins/stack_connectors/server/connector_types/security/tines/tines.ts:
```diff
@@ -0,0 +1,177 @@
+    body,
+  }: TinesRunActionParams): Promise<TinesRunActionResponse> {
+    if (!webhook && !webhookUrl) {
+      throw Error('Invalid subActionsParams: [webhook] or [webhookUrl] expected but got none');
```
This error was needed because of a limitation of the schema validation. This action parameter needs an object with: 
```
body & (webhook | webhookUrl)
```
I haven't found a way to define an object with "one of two" properties, so I added a _maybe_ on both and created this error. The schema validation errors are not translated, so I did not translate this one, I followed the same message format.
This error can not happen under normal conditions, however, a custom API request with only the body can make it to this point, so I added this error only to prevent the request be sent with no url.

semd commented on x-pack/plugins/stack_connectors/public/connector_types/security/tines/tines_params.tsx:
```diff
@@ -0,0 +1,330 @@
+    }
+  }, [toasts, storiesError, webhooksError]);
+
+  const showFallbackFrom = useMemo<'Story' | 'Webhook' | 'any' | null>(() => {
```
It is not an API failure, the response is successful, it's just that we know we hit the result limit (there are more results that are not included in the response data), so we show a fallback/alternative way to enter the parameter.
Also, the fallback field is also shown if the `webhookUrl` value is stored in the connector, even if we haven't hit the limit (the `any` value).
The _From_ at the end refers to the reason why we show the fallback, we need to know that as well.
I haven't found a better name. maybe `showAlternativeInputFrom`?

jonathan-buttner replied:
Ah ok. Totally a `nit` so you don't need to change it but maybe `fallbackReason`?

Maybe we could switch `any` to `urlSpecified` or something?

semd commented on x-pack/plugins/stack_connectors/public/connector_types/security/tines/tines.ts:
```diff
@@ -0,0 +1,108 @@
+      } else {
+        if (!subActionParams?.webhook?.storyId) {
+          errors.story.push(translations.STORY_REQUIRED);
+        } else {
```
If the `story` is not defined the webhook will never be defined. Also, the webhook selector is disabled, it does not make much sense to show an error that the user can not fix. Once the story is defined the "webhook required" error is checked and the webhook selector is enabled so the user can choose one. I think this way makes more sense.

semd commented on x-pack/plugins/stack_connectors/public/connector_types/security/tines/tines.ts:
```diff
@@ -0,0 +1,108 @@
+          const parsedUrl = new URL(subActionParams.webhookUrl);
+          if (parsedUrl.protocol !== 'https:') {
+            errors.webhookUrl.push(translations.INVALID_PROTOCOL_WEBHOOK_URL);
+          } else if (!parsedUrl.hostname.endsWith('.tines.com')) {
```
Haven't thought about it, do you think is better not to check this one?

semd commented on x-pack/plugins/triggers_actions_ui/public/application/hooks/use_sub_action.test.tsx:
```diff
@@ -131,27 +128,37 @@ describe('useSubAction', () => {
+      abortSpy.mockRestore();
+      return firstAbortCtrl;
+    });
+    mockHttpPost.mockImplementationOnce(
```
It is not exactly the same, I use `setTimeout` inside the promise so the response is resolved at the end of the event loop, this way the `rerender` call happens before the first request has finished. 
Using `mockResolvedValue` the response would resolve immediately, the test checks would also pass, but we would not be testing it properly, in the following test case it is more obvious.

semd commented on x-pack/plugins/stack_connectors/server/connector_types/security/tines/tines.test.ts:
```diff
@@ -0,0 +1,226 @@
+      expect(mockRequest).toBeCalledTimes(1);
+      const [[requestArgs]] = mockRequest.mock.calls;
+
+      expect(requestArgs).toEqual(storiesGetRequestExpected);
```
yep, this was left from when we had to check the parameters from the first and second calls (pagination) separately. Fixed

semd commented on x-pack/plugins/stack_connectors/server/connector_types/security/tines/tines.ts:
```diff
@@ -0,0 +1,175 @@
+    if (error.response.status === 401) {
+      return 'Unauthorized API Error';
+    }
+    return `API Error: (${error.response?.status}) ${error.response?.statusText}`;
```
I asked for an error reference and they don't have it. They say we should base our error control on the status code only, there's no additional information. So I use the `statusText` standard to give some information to the user.

semd commented on x-pack/plugins/stack_connectors/server/connector_types/security/tines/tines.ts:
```diff
@@ -0,0 +1,175 @@
+      throw Error('Invalid subActionsParams: [webhook] or [webhookUrl] expected but got none');
+    }
+    const response = await this.request({
+      url: webhookUrl ? webhookUrl : this.urls.getRunWebhookURL(webhook!),
```
Yes, it is needed. TS doesn't understand that the error prevents this situation, I should add a return statement after the error to make TS understand. What do you prefer?

semd commented on x-pack/plugins/stack_connectors/public/connector_types/security/tines/tines_params.test.tsx:
```diff
@@ -0,0 +1,537 @@
+  const original = jest.requireActual(triggersActionsPath);
+  return {
+    ...original,
+    useSubAction: (params: UseSubActionParams<unknown>) => mockUseSubAction(params),
```
This is a unit test for the form, I don't think we should test the `useSubAction` hook along with it. The `useSubAction` hook has its own unit test, and there's also an integration test that checks that all the UI components used in this form work properly together.

We should not have to come and do changes to this unit test unless there's some change in the component that it is testing. Otherwise, it would mean that this is not a unit but an integration test, and we already have one.

I mean, if we don't mock the `useSubAction` in the unit tests of the components that use it, a change in the internal logic of this hook may cause all those tests to fail, even if nothing changed from the outside and everything is working fine, introducing false negatives and making us go and fix all those tests unnecessarily.
This is not how a unit test should work. they should have only one responsibility, test only one thing, and we should avoid coupling them with other things.

cnasikas replied:
We do not test the `useSubAction` by not mocking it. We test how our component interacts with the hook. Imagine having a hook with one `useEffect`. By mocking the hook you do not really test how your component behaves with the extra `useEffect` because you mocked it. Also, by mocking it we do not test how our component behaves with async operations like fetching etc.

> I mean, if we don't mock the useSubAction in the unit tests of the components that use it, a change in the internal logic of this hook may cause all those tests to fail, even if nothing changed from the outside and everything is working fine, introducing false negatives and making us go and fix all those tests unnecessarily.

This is more of a reason to not mock the hook. If we do a change in the hook and the tests of the component break then it means that we introduce a bug in the hook. By mocking it you will not be aware of that. A change in the hook should pass the tests either way.

>This is not how a unit test should work. they should have only one responsibility, test only one thing, and we should avoid coupling them with other things.?

This is correct. Not mocking the hooks does not mean that you turn a unit test into an integration one. It is still a unit test with only one responsibility, testing the component. And the component is coupled with all hooks it uses.

I know we tend to mock the hooks a lot, especially the ones that fetch, but I think we should stop doing it and mock only what is necessary.  Nevertheless, if you prefer to mock it I am fine. I just wanted to mention it. Some interesting articles about the topic: https://kentcdodds.com/blog/stop-mocking-fetch, https://tkdodo.eu/blog/testing-react-query#mocking-network-requests

semd commented on x-pack/plugins/stack_connectors/public/connector_types/security/tines/tines_params.tsx:
```diff
@@ -0,0 +1,330 @@
+    WebhookOption | null | undefined
+  >();
+
+  const isTesting = useMemo(() => !messageVariables?.length, [messageVariables]);
```
nice! it didn't exist when I did this. much better 👍

semd commented on x-pack/plugins/stack_connectors/public/connector_types/security/tines/tines_params.tsx:
```diff
@@ -0,0 +1,330 @@
+  const storiesOptions = useMemo(() => stories?.map(createOption) ?? [], [stories]);
+  const webhooksOptions = useMemo(() => webhooks?.map(createOption) ?? [], [webhooks]);
+
+  useEffect(() => {
```
Nope. If we put this one outside the `useEffect`, and one request had an error, any component update will trigger the error toast again and again:

![errors](https://user-images.githubusercontent.com/17747913/201124185-57513f9c-3999-4be2-b506-7802c97c0558.png)

semd commented on x-pack/plugins/stack_connectors/public/connector_types/security/tines/tines_params.tsx:
```diff
@@ -0,0 +1,330 @@
+  ]);
+
+  useEffect(() => {
+    if (selectedStoryOption === undefined && webhook?.storyId && stories) {
```
it is not a good practice to introduce side effects inside `useMemo`, that's what `useEffect` is for.

cnasikas replied:
Sorry I did not mean to set the state inside the `useMemo`. I meant that you do not need to have a state for that. You can derive it from the `webhook`. Btw, setting a state is not a "side effect" and it should not be used with a `useEffect` (using it with `editSubActionParams` it is fine). This is a very interesting article about using `useEffect`: https://beta.reactjs.org/learn/you-might-not-need-an-effect

semd replied:
true, I wanted to say "set state" not "side effect", but neither are recomended to do inside `useMemo`. I don't understand what state you think is not needed, have you tested your idea? the component is pretty complex.

semd commented on x-pack/plugins/stack_connectors/public/connector_types/security/tines/tines_params.tsx:
```diff
@@ -0,0 +1,330 @@
+  }, [selectedStoryOption, webhook?.storyId, stories, toasts, editSubActionParams]);
+
+  useEffect(() => {
+    if (selectedWebhookOption === undefined && webhook?.id && webhooks) {
```
same

semd commented on x-pack/plugins/stack_connectors/public/connector_types/security/tines/tines_params.tsx:
```diff
@@ -0,0 +1,330 @@
+  const webhooksOptions = useMemo(() => webhooks?.map(createOption) ?? [], [webhooks]);
+
+  useEffect(() => {
+    if (storiesError) {
```
We can do this improvement later. This PR is complex enough

cnasikas replied:
I agree, no need for this PR.

cnasikas commented on x-pack/plugins/stack_connectors/server/connector_types/security/tines/render.ts:
```diff
@@ -0,0 +1,33 @@
+  if (params?.subAction !== SUB_ACTION.RUN) return params;
+
+  // Remove the "kibana" entry from all alerts to reduce weight, the same data can be found in other parts of the alert object.
+  const alerts = (variables.context as Context).alerts.map(({ kibana, ...alert }) => alert);
```
In Kibana we can always migrate SO when changing the schema. I am not sure what changes the alert-as-data initiative will do so I just wanted to raise awareness of a pontentionally issue. Maybe I am wrong and we do not need to do anything. @ymao1 Any input on this?

ymao1 replied:
Thanks for bringing this up @cnasikas!

This is an accurate assessment as right now, each rule type defines its own context variables that it populates to pass through to actions/notifications. If a rule type has `context.variableA` and changes it to `context.variableB`, we don't have a process for migrating the mustache templates that are stored in the rule saved object and the actions that use `context.variableA` would render empty data.

That said, this is assuming a very detection rule specific context. Maybe that's ok with this PR since it'll just be enabled for detection rules currently but I would like to see some error handling for the case where `context.alerts` is not defined (as is the case for every rule type other than detection rules).

Also, I would agree with @cnasikas that it an allowlist makes sense. At a minimum, I would remove the `kibana.alert.rule.*` keys since we are evaluating the necessity of keeping them in AAD and the information in them is stored elsewhere (in `params` and at the top level `rule` object).

ymao1 replied:
Oh I see! That is even better!

semd commented:
@elastic/mlr-docs could you please take a look?

jonathan-buttner reviewed (approved):
Great work @semd !

semd commented:
@elasticmachine merge upstream
